apps/music/js/metadata/flac.js

```diff
@@ -36,37 +37,64 @@ var FLACMetadata = (function() {
     // First four bytes are "fLaC" or we wouldn't be here.
     blobview.seek(4);
 
-    metadata.tag_format = 'flac';
-    return findVorbisCommentBlock(blobview).then(function(block) {
-      if (!block) {
-        return metadata;
+    return new Promise(function(resolve, reject) {
+      metadata.tag_format = 'flac';
+      var has_vorbis_comment = false;
+      var has_picture = false;
+
+      function processBlock(block) {
+        if (!block) {
+          resolve(metadata);
+          return;
+        }
+
+        if (block.block_type === 4) {
+          readAllComments(block.view, metadata);
+          has_vorbis_comment = true;
+
+        } else if (block.block_type === 6) {
+          LazyLoader.load('js/metadata/vorbis_picture.js', function() {
+            metadata.picture = VorbisPictureComment.readPicFrame(block.view);
+            metadata.picture.start += block.view.viewOffset;
+            metadata.picture.end += block.view.viewOffset;
+          });
+          has_picture = true;
+
+        }
+
+        if (!has_vorbis_comment || !has_picture) {
+          try {
+            block.view.advance(block.length - block.view.index);
+            findMetadataBlocks(block.view, processBlock);
+          }
+          catch(e) {
+            console.error(e.message);
+            resolve(metadata);
+          }
+        }
+        resolve(metadata);
       }
-      return readAllComments(block.view, metadata);
+
+      findMetadataBlocks(blobview, processBlock);
     });
   }
 
   /**
-   * Step over metadata blocks until we find the Vorbis comment block.
+   * Step over metadata blocks until we find the proper metadata block.
    *
    * @param {BlobView} blobview The BlobView for the file.
-   * @return {Promise} A promise resolving to an object describing the metadata
-   *   block. See readMetadataBlockHeader for more details.
+   * @param {function} callback The callback to process the block with.
+   *   See readMetadataBlockHeader for more details.
    */
-  function findVorbisCommentBlock(blobview) {
-    return readMetadataBlockHeader(blobview).then(function(block) {
-      // XXX: Support album art.
-      // See: http://flac.sourceforge.net/format.html#metadata_block_picture
-
-      // Did we find a Vorbis comment block yet?
-      if (block.block_type === 4) {
-        return block;
-      } else if (block.last) {
-        return null;
-      } else {
-        block.view.advance(block.length);
-        return findVorbisCommentBlock(block.view);
-      }
-    });
+  function findMetadataBlocks(blobview, callback) {
+    readMetadataBlockHeader(
+      blobview, function(block, err) {
+        if (err || block.last) {
+          callback(null);
+        } else {
+          callback(block);
```

Comment:
To elaborate a little more (and to explain why I like `readMetadataBlockHeader` the way it was), here's how I'd write this function. This also fixes an off-by-one bug that's my fault (We still need to process the last block, but then we need to stop *after* that one):

```
function findMetadataBlocks(blobview, callback) {
  readMetadataBlockHeader(blobview).then(function(block) {
    callback(block);
    if (block.last) {
      callback(null);
    } else {
      block.view.advance(block.length - block.view.index);
      findMetadataBlocks(block.view, callback);
    }
  }).catch(function(err) {
    console.error(err);
    callback(null);
  });
}
```

This has the nice side-effect that if any errors are thrown during our `then`, the `catch` will automatically pick them up, so we don't need to worry about adding a try block around `block.view.advance(...)`.